lib/charms/istio_k8s/v0/istio_ingress_config.py

```diff
@@ -0,0 +1,285 @@
+        """Publish OAuth configuration data to all related applications.
+
+        Args:
+            oauth_service_name: The OAuth service name.
+            oauth_port: The port number for the OAuth service.
+        """
+        data = ProviderIngressConfigData(
+            oauth_service_name=oauth_service_name,
+            oauth_port=oauth_port,
+            model_name=self._model_name,
+        ).model_dump(mode="json", by_alias=True, exclude_defaults=True, round_trip=True)
+
+        for relation in self.relations:
+            databag = relation.data[self._app]
+            databag.update(data)
+            log.debug("Published provider data: %s to relation: %s", data, relation)
+
+    def get_oauth_provider_name(self) -> Optional[str]:
+        """Fetch the OAuth provider name generated by the requirer for this provider.
+
+        Returns:
+            The generated OAuth provider name if available, else None.
+        """
+        for relation in self.relations:
+            # Access the data published by the requirer.
+            raw_data = getattr(relation, "data", {}).get(relation.app)
+            if raw_data:
+                try:
+                    req_data = RequirerIngressConfigData.model_validate(raw_data)
+                    if self._app.name == req_data.ingress_provider_name:
+                        return req_data.oauth_provider_name
+                except Exception as e:
+                    log.debug("Failed to validate requirer data: %s", e)
+                    continue
+        return None
+
+    def is_requirer_ready(self) -> bool:
+        """Guard to check if the generated OAuth provider name is present.
+
+        Returns:
+            True if the OAuth provider name has been published by the requirer.
+        """
+        return self.get_oauth_provider_name() is not None
+
+
+class IngressConfigRequirer:
+    """Requirer side wrapper for the istio-ingress-config relation.
+
+    The requirer generates and publishes a unique OAuth provider name
+    for a connected ingress charm. It can also check that the provider has published
+    its required OAuth service configuration.
+    """
+
+    def __init__(
+        self,
+        relation_mapping: RelationMapping,
+        app: Application,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        """Initialize the IngressConfigRequirer.
+
+        Args:
+            relation_mapping: The charm's RelationMapping (typically self.model.relations).
+            app: This application.
+            relation_name: The name of the relation.
+        """
+        self._charm_relation_mapping = relation_mapping
+        self._app = app
+        self._relation_name = relation_name
+
+    @property
+    def relations(self):
+        """Return the relation instances for the monitored relation."""
+        return self._charm_relation_mapping.get(self._relation_name, ())
```

Comment:
I believe type checking/Pietro reviewing will eventually lead you to this :D  Ask me how I know ...

```suggestion
    def relations(self) -> List[Relation]:
        """Return the relation instances for the monitored relation."""
        return self._charm_relation_mapping.get(self._relation_name, [])
```